File: src/lib/api/adapters/web/checkpoints.ts
import { apiClient } from '../../client';
import type { CheckpointMeta, CheckpointDiffResult, CheckpointOperations, GitDiff } from '../../types';

export class WebCheckpointAdapter implements CheckpointOperations {
  async list(namespace: string): Promise<CheckpointMeta[]> {
    const params = new URLSearchParams({ ns: namespace });
    const res = await apiClient.get<{ checkpoints: any[] }>(`/api/checkpoints?${params}`);
    return (res.checkpoints || []).map((cp: any) => ({
      id: cp.id,
      namespace: cp.namespace,
      title: cp.title,
      bullets: cp.bullets,
      createdAt: cp.createdAt ?? cp.created_at ?? cp.createdAt,
      author: cp.author,
    } as CheckpointMeta));
  }
  async create(namespace: string, metadata?: { reason?: string; actor?: string; paths?: string[] }): Promise<CheckpointMeta> {
    const params = new URLSearchParams({ ns: namespace });
    const res = await apiClient.post<{ checkpoint: CheckpointMeta }>(`/api/checkpoints?${params}`, metadata || {});
    return res.checkpoint;
  }
  async diff(namespace: string, baseId: string, targetRef: 'HEAD' | 'WORKTREE' | string = 'HEAD'): Promise<CheckpointDiffResult> {
    const params = new URLSearchParams({ ns: namespace, base: baseId, target: String(targetRef) });
    const res = await apiClient.get<{ diff: GitDiff }>(`/api/checkpoints/diff?${params}`);
    return { diff: res.diff };
  }
  async restore(namespace: string, id: string): Promise<{ result: string; commitId?: string }> {
    const params = new URLSearchParams({ ns: namespace, id });
    return await apiClient.post<{ result: string; commitId?: string }>(`/api/checkpoints/restore?${params}`, {});
  }
  async publish(namespace: string): Promise<{ result: string; mergeCommitId?: string }> {
    const params = new URLSearchParams({ ns: namespace });
    return await apiClient.post<{ result: string; mergeCommitId?: string }>(`/api/checkpoints/publish?${params}`, {});
  }
}

File: backend/src/transport/routes/checkpoints.rs
use axum::{
    extract::{Query, State},
    http::StatusCode,
    response::Json,
    routing::{get, post},
    Router,
};
use serde::{Deserialize, Serialize};
use std::sync::Arc;

use crate::svc::git_service::GitService;
use crate::repo::git_repository::{CheckpointMeta, GitDiff};

#[derive(Debug, Deserialize)]
pub struct ListQuery { pub ns: String, pub max: Option<usize> }

#[derive(Debug, Deserialize)]
pub struct CreateQuery { pub ns: String }

#[derive(Debug, Deserialize)]
pub struct DiffQuery { pub ns: String, pub base: String, pub target: String }

#[derive(Debug, Deserialize)]
pub struct RestoreQuery { pub ns: String, pub id: String }

#[derive(Debug, Deserialize)]
pub struct PublishQuery { pub ns: String }

#[derive(Debug, Deserialize)]
pub struct CreateBody { pub title: Option<String>, pub reason: Option<String>, pub actor: Option<String> }

#[derive(Debug, Deserialize)]
pub struct RestoreBody { pub message: Option<String> }

#[derive(Debug, Serialize)]
pub struct ListResponse { pub checkpoints: Vec<CheckpointMeta> }

#[derive(Debug, Serialize)]
pub struct CreateResponse { pub checkpoint: CheckpointMeta }

#[derive(Debug, Serialize)]
pub struct DiffResponse { pub diff: GitDiff }

#[derive(Debug, Serialize)]
pub struct RestoreResponse { pub result: String, #[serde(rename = "commitId")] pub commit_id: Option<String> }

#[derive(Debug, Serialize)]
pub struct PublishResponse { pub result: String, #[serde(rename = "mergeCommitId")] pub merge_commit_id: Option<String> }

#[derive(Debug, Serialize)]
pub struct ErrorResponse { pub error: String }

pub fn checkpoints_router() -> Router<Arc<GitService>> {
    Router::new()
        .route("/", get(list_checkpoints).post(create_checkpoint))
        .route("/diff", get(diff_checkpoint))
        .route("/restore", post(restore_checkpoint))
        .route("/publish", post(publish_checkpoints))
}

async fn list_checkpoints(
    State(svc): State<Arc<GitService>>,
    Query(q): Query<ListQuery>,
) -> Result<Json<ListResponse>, (StatusCode, Json<ErrorResponse>)> {
    match svc.checkpoints_list(&q.ns, q.max).await {
        Ok(items) => Ok(Json(ListResponse { checkpoints: items })),
        Err(e) => Err((StatusCode::INTERNAL_SERVER_ERROR, Json(ErrorResponse { error: e.to_string() }))),
    }
}

async fn create_checkpoint(
    State(svc): State<Arc<GitService>>,
    Query(q): Query<CreateQuery>,
    Json(body): Json<CreateBody>,
) -> Result<Json<CreateResponse>, (StatusCode, Json<ErrorResponse>)> {
    match svc.checkpoints_create(&q.ns, body.title).await {
        Ok(cp) => Ok(Json(CreateResponse { checkpoint: cp })),
        Err(e) => Err((StatusCode::INTERNAL_SERVER_ERROR, Json(ErrorResponse { error: e.to_string() }))),
    }
}

async fn diff_checkpoint(
    State(svc): State<Arc<GitService>>,
    Query(q): Query<DiffQuery>,
) -> Result<Json<DiffResponse>, (StatusCode, Json<ErrorResponse>)> {
    match svc.checkpoints_diff(&q.ns, &q.base, &q.target).await {
        Ok(diff) => Ok(Json(DiffResponse { diff })),
        Err(e) => Err((StatusCode::INTERNAL_SERVER_ERROR, Json(ErrorResponse { error: e.to_string() }))),
    }
}

async fn restore_checkpoint(
    State(svc): State<Arc<GitService>>,
    Query(q): Query<RestoreQuery>,
    Json(body): Json<RestoreBody>,
) -> Result<Json<RestoreResponse>, (StatusCode, Json<ErrorResponse>)> {
    match svc.checkpoints_restore(&q.ns, &q.id, body.message).await {
        Ok(commit) => Ok(Json(RestoreResponse { result: "ok".into(), commit_id: Some(commit.sha) })),
        Err(e) => Err((StatusCode::INTERNAL_SERVER_ERROR, Json(ErrorResponse { error: e.to_string() }))),
    }
}

async fn publish_checkpoints(
    State(svc): State<Arc<GitService>>,
    Query(q): Query<PublishQuery>,
) -> Result<Json<PublishResponse>, (StatusCode, Json<ErrorResponse>)> {
    match svc.checkpoints_publish(&q.ns).await {
        Ok(oid) => Ok(Json(PublishResponse { result: "ok".into(), merge_commit_id: Some(oid) })),
        Err(e) => Err((StatusCode::INTERNAL_SERVER_ERROR, Json(ErrorResponse { error: e.to_string() }))),
    }
}
